src/frontegg-auth/src/auth.rs

```diff
@@ -314,206 +274,314 @@ impl Authentication {
+        mz_ore::task::spawn(|| name, {
+            let inner = Arc::clone(self);
+            async move {
+                tracing::debug!(?password.client_id, "starting refresh task");
+                let gauge = inner.metrics.refresh_tasks_active.with_label_values(&[]);
+                gauge.inc();
+
+                loop {
+                    let valid_for = f64::cast_lossy(claims.exp - inner.now.as_secs());
+                    // If we have no outstanding handling to this session, but a handle was dropped
+                    // within this window, then we'll still refresh.
+                    let drop_window = u64::cast_lossy(valid_for * inner.refresh_drop_factor).max(1);
+                    // Scale the validity duration by 0.8. The Frontegg Python
+                    // SDK scales the expires_in this way.
+                    //
+                    // <https://github.com/frontegg/python-sdk/blob/840f8318aced35cea6a41d83270597edfceb4019/frontegg/common/frontegg_authenticator.py#L45>
+                    let valid_for = u64::cast_lossy(valid_for * 0.8);
+
+                    if valid_for < 60 {
+                        warn!(%valid_for, "unexpectedly low token validity");
+                    }
 
-            let current_response = client
-                .refresh_token(&refresh_url, &latest_response.refresh_token, &metrics)
-                .await;
+                    tracing::debug!(%valid_for, %drop_window, "waiting for token validity period");
+
+                    // Wait out validity duration.
+                    time::sleep(Duration::from_secs(valid_for)).await;
+
+                    // Check to see if all external metadata receivers have gone away, or if a
+                    // session created with this password was recently dropped. If no one is
+                    // listening nor any recent handles were dropped we can clean up the session.
+                    let receiver_count = external_metadata_tx.receiver_count();
+                    let last_drop = inner.last_dropped_session(&password);
+                    let recent_drop = last_drop
+                        .map(|dropped_at| dropped_at.elapsed() <= Duration::from_secs(drop_window))
+                        .unwrap_or(false);
+                    if receiver_count == 0 && !recent_drop {
+                        tracing::debug!(
+                            ?last_drop,
+                            ?password.client_id,
+                            "all listeners have dropped and none of them were recent!"
+                        );
+                        break;
+                    }
 
-            // If we failed to refresh, then close the channel.
-            let current_response = match current_response {
-                Ok(resp) => {
-                    tracing::debug!(?id, "refresh successful");
-                    resp
+                    // TODO(parkmycar): Allocating the labels here feels bad.
```

Comment:
I agree, and think changing this needs to be done before merging because of the unpredictable cardinality. Can these be separate metrics?

Reply:
Very true! Changed the count of receivers to just are there >0, so we no longer need to allocate a String, and the cardinality is fixed